File: 2025/08 August/18 Find H-Index/main.py
# Link: https://www.geeksforgeeks.org/problems/find-h-index--165609/1


class Solution:
    def hIndex(self, citations):
        n = len(citations)
        freq = [0] * (n + 1)

        for citation in citations:
            if citation >= n:
                freq[n] += 1
            else:
                freq[citation] += 1

        idx = n

        s = freq[n]
        while s < idx:
            idx -= 1
            s += freq[idx]

        return idx


if __name__ == "__main__":
    sol = Solution()
    citations = [3, 0, 5, 3, 0]
    result = sol.hIndex(citations)
    print(f"The H-Index is: {result}")

    citations = [5, 1, 2, 4, 1]
    result = sol.hIndex(citations)
    print(f"The H-Index is: {result}")

    citations = [0, 0]
    result = sol.hIndex(citations)
    print(f"The H-Index is: {result}")

File: 2025/08 August/18 Find H-Index/main.cpp
// Link: https://www.geeksforgeeks.org/problems/find-h-index--165609/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int hIndex(vector<int> &citations)
    {
        int n = citations.size();
        vector<int> freq(n + 1);

        for (int i = 0; i < n; i++)
        {
            if (citations[i] >= n)
                freq[n] += 1;
            else
                freq[citations[i]] += 1;
        }

        int idx = n;

        int s = freq[n];
        while (s < idx)
        {
            idx--;
            s += freq[idx];
        }

        return idx;
    }
};

int main()
{
    Solution sol;
    vector<int> citations = {3, 0, 5, 3, 0};
    int result = sol.hIndex(citations);
    cout << "H-Index: " << result << endl;

    citations = {5, 1, 2, 4, 1};
    result = sol.hIndex(citations);
    cout << "H-Index: " << result << endl;

    citations = {0, 0};
    result = sol.hIndex(citations);
    cout << "H-Index: " << result << endl;
    return 0;
}